### PR TITLE
Add cloud-agnostic CloudClient interface

### PR DESCRIPTION
**Pull request by mbarnes**
This is the completion of @lisa's https://github.com/lisa/cloud-ingress-operator/tree/osd-3566 branch.

This was a collective Team Rocket effort, so apologies for the messy commit history.

All controllers have been updated to use the new `CloudClient` interface.

The utility functions proved hardest to wrangle.  I've introduced a new `pkg/utils` subpackage which conceptually sits "below" the  `pkg/controller` and `pkg/cloudclient` layers so as to avoid import loops.  This is where `GetInfrastructureObject` and most of the utilities that rely on it ended up.

### CODE REVIEW

**Comment by lisa** [comment]
/hold

(for review and eventual squashing)

**Review by lisa — CHANGES_REQUESTED** [review]
What is the status of testing in-cluster here?

Love to see this getting to PR state.

Can we consolidate commits as per Slack convo?

File: pkg/cloudclient/aws/private.go [hunk 1]
```diff
@@ -0,0 +1,831 @@
...
+		err = updateAWSLBList(kclient, lbList, newLBList, machine, providerSpecDecoded)
+		if err != nil {
+			log.Error(err, "Error updating LB list for machine", "machine", machine.Name)
+			return err
+		}
+	}
+	return nil
+}
+
+// getAWSDecodedProviderSpec casts the spec.providerSpec of a kubernetes machine
```
[the remark this answers is not in the document]
Comment (mbarnes) [reply pkg/cloudclient/aws/private.go:112]: :heavy_check_mark:

File: pkg/cloudclient/aws/private.go [hunk 2]
```diff
@@ -0,0 +1,831 @@
...
+	listOptions := []client.ListOption{
+		client.InNamespace("openshift-machine-api"),
+		client.MatchingLabels{masterMachineLabel: "master"},
+	}
+	err := kclient.List(context.TODO(), machineList, listOptions...)
+	if err != nil {
+		return nil, err
+	}
+	return machineList, nil
+}
```
[the remark this answers is not in the document]
Comment (mbarnes) [reply pkg/cloudclient/aws/private.go:369]: I think it got dragged in with `GetMasterNodeSubnets`, but I'll split it back out.

[the remark this answers is not in the document]
Comment (mbarnes) [reply pkg/cloudclient/aws/private.go:369]: :heavy_check_mark:

File: pkg/cloudclient/aws/private.go [hunk 3]
```diff
@@ -0,0 +1,831 @@
...
+			return err
+		}
+	}
+	return nil
+}
+
+// AddAWSLBToMasterMachines adds a Load Balancer (with name elbName) to the
+// spec.providerSpec.value.loadBalancers list for each of the master machine
+// objects in a cluster
+func AddAWSLBToMasterMachines(kclient client.Client, elbName string, masterNodes *machineapi.MachineList) error {
```
[the remark this answers is not in the document]
Comment (mbarnes) [reply pkg/cloudclient/aws/private.go:84]: :heavy_check_mark:

File: pkg/cloudclient/aws/private_test.go [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+func (m mockCreateLoadBalancer) CreateLoadBalancer(i *elbv2.CreateLoadBalancerInput) (*elbv2.CreateLoadBalancerOutput, error) {
+	var e awserr.Error
+	if m.ErrResp != "" {
+		e = awserr.New(m.ErrResp, m.ErrResp, fmt.Errorf("Error raised by test"))
+	}
+	return &m.Resp, e
+}
+
+func TestCreateNetworkLoadBalancer(t *testing.T) {
+	// TODO: Fill this in
```
[the remark this answers is not in the document]
Comment (mbarnes) [reply pkg/cloudclient/aws/private_test.go:434]: Looks like that comment's been there awhile.  I see a test case so I assume it's completed.

[the remark this answers is not in the document]
Comment (mbarnes) [reply pkg/cloudclient/aws/private_test.go:434]: :heavy_check_mark:

**Comment by mbarnes** [comment]
I've cleaned up the commit history so it shows a more concise and straight-forward progression, while retaining proper authorship.  Also addressed all review comments so far.  Force-pushing the cleaner commit list.

> What is the status of testing in-cluster here?
Everything seemed to be working before I started messing with commits.  I'll re-verify.

File: pkg/controller/publishingstrategy/publishingstrategy_controller.go [hunk 5]
```diff
@@ -182,193 +179,32 @@ func (r *ReconcilePublishingStrategy) Reconcile(request reconcile.Request) (reco
...
-
-		// upsert resource record to change api.<clusterName> from external NLB to internal NLB
-		err = awsClient.UpsertARecord(pubDomainName+".", intDNSName, intHostedZoneID, apiDNSName, comment, false)
+		err := cloudClient.SetDefaultAPIPrivate(context.TODO(), r.client, instance)
 		if err != nil {
 			log.Error(err, "Error updating api.<clusterName> alias to internal NLB")
 			return reconcile.Result{}, err
 		}
-		log.Info(fmt.Sprintf("%s successful", comment))
+		log.Info("Update api.<clusterName> alias to internal NLB successful")
```
[the remark this answers is not in the document]
Comment (mbarnes) [reply pkg/controller/publishingstrategy/publishingstrategy_controller.go:195]: :heavy_check_mark:

File: pkg/controller/publishingstrategy/publishingstrategy_controller.go [hunk 6]
```diff
@@ -182,193 +179,32 @@ func (r *ReconcilePublishingStrategy) Reconcile(request reconcile.Request) (reco
...
 			log.Error(err, "Error updating api.<clusterName> alias to internal NLB")
 			return reconcile.Result{}, err
 		}
-		log.Info(fmt.Sprintf("%s successful ", comment))
-		err = utils.AddAWSLBToMasterMachines(r.client, extNLBName, masterList)
-		if err != nil {
-			log.Error(err, "Error adding new LB to master machines' providerSpecs")
-			return reconcile.Result{}, err
-		}
+		log.Info("Update api.<clusterName> alias to internal NLB successful")
```
[the remark this answers is not in the document]
Comment (mbarnes) [reply pkg/controller/publishingstrategy/publishingstrategy_controller.go:207]: :heavy_check_mark:

**Comment by lisa** [comment]
It is with great pleasure that I get to be the one to do this:

/lgtm

:shipit:   🎉 👍 

Great work to everyone involved with this port.


@mbarnes Please `/unhold` when ready to merge.

**Review by lisa — APPROVED** [review]

**Pull request merged**